bitwarden_access_manager.py

```diff
@@ -0,0 +1,725 @@
+type = "member"
+groups = ["group1"]
+
+[[member]]
+member_id = "856cba2d-cae1-40e7-96cc-afaf00a8a4cb"
+member_name = "yunkel"
+email = "yunkel68@hotmail.fr"
+type = "owner"
+access_all = true
+groups = ["group1", "group2"]
+
+[[group]]
+group_id = "c6a13b93-edc1-4c3b-9fc5-afaf00a8d33f"
+group_name = "group1"
+access_all = true
+
+[[group]]
+group_id = "39b48ab2-81fd-40eb-87e9-afb0000110f3"
+group_name = "group2"
+
+[[collection]]
+collection_id = "50351c20-55b4-4ee8-bbe0-afaf00a8f25d"
+external_id = "collection1"
+member_access = [
+  { member_name = "yan", access = "write"},
+]
+
+group_access = [
+  { group_name = "group1", access = "readonly"},
+  { group_name = "group2", access = "write" },
+]
+
+[[collection]]
+collection_id = "8e69ce49-85ae-4e09-a52c-afaf00a90a3f"
+external_id = ""
+member_access = [
+  { member_name = "yan", access = "write" },
+]
+group_access = [
+  { group_name = "group1", access = "readonly" },
+]
+"""
+
+from __future__ import annotations
+from collections import defaultdict
+from dataclasses import dataclass
+from difflib import SequenceMatcher
+from enum import Enum
+from http.client import HTTPSConnection, HTTPResponse
+
+import json
+import os
+import requests
+import sys
+import tomllib
+
+from typing import (
+    Any,
+    Dict,
+    List,
+    Generic,
+    Tuple,
+    NamedTuple,
+    Set,
+    Iterable,
+    Protocol,
+    TypeVar,
+)
+
+
+class MemberType(Enum):
+    OWNER = 0
+    ADMIN = 1
+    USER = 2
+    MANAGER = 3
+    CUSTOM = 4
+
+
+class GroupAccess(Enum):
+    READONLY = 0
+    WRITE = 1
+
+
+class Member(NamedTuple):
+    id: str
+    name: str
+    email: str
+    type: MemberType
+    access_all: bool
+    groups: Tuple[str, ...]
+
+    def get_id(self) -> str:
+        return self.id
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> Member:
+        access_all: bool = False
+        groups: Tuple[str, ...] = tuple()
+
+        if "access_all" in data:
+            access_all = data["access_all"]
+        if "groups" in data:
+            groups = data["groups"]
+            groups = tuple(sorted(data["groups"]))
+        return Member(
+            id=data["member_id"],
+            name=data["member_name"],
+            email=data["email"],
+            type=MemberType[data["type"].upper()],
+            access_all=access_all,
+            groups=groups,
+        )
+
+    def format_toml(self) -> str:
+        result = (
+            "[[member]]\n"
+            f'member_id = "{self.id}"\n'
+            f'member_name = "{self.name}"\n'
+            f'email = "{self.email}"\n'
+            f'type = "{self.type.name.lower()}\n"'
+            f"access_all = {str(self.access_all).lower()}\n"
+        )
+
+        groups = self.groups or ()
+        if len(groups) > 0:
+            groups_str = ", ".join(f'"{g}"' for g in sorted(groups))
+            result = result + "groups = [ " + groups_str + " ]"
+        else:
+            result = result + "groups = []"
```

Comment:
Why do we need to convert `g` into a json string here ?

Reply:
If we just surround `g` by quotes, then if the string `g` contains a `"` or a `\`, it will produce invalid toml. If we use `json.dumps` to format a json string literal, then it will escape `"` and `\` to `\"` and `\\`. This works because [toml has the same escaping rules as json](https://toml.io/en/v1.0.0#string).